Advanced-Audio-Driver-Kotlin/app/src/main/java/com/tokbox/sample/advancedaudiodriver/MainActivity.kt

```diff
@@ -0,0 +1,224 @@
+            Log.d(TAG, "onStreamDestroyed: Own stream " + stream.streamId + " destroyed")
+        }
+
+        override fun onError(publisherKit: PublisherKit, opentokError: OpentokError) {
+            finishWithMessage("PublisherKit error: " + opentokError.message)
+        }
+    }
+    private val sessionListener: SessionListener = object : SessionListener {
+        override fun onConnected(session: Session) {
+            Log.d(TAG, "onConnected: Connected to session " + session.sessionId)
+
+            publisher = Publisher.Builder(this@MainActivity).build().also {
+                it.setPublisherListener(publisherListener)
+                it.setStyle(BaseVideoRenderer.STYLE_VIDEO_SCALE, BaseVideoRenderer.STYLE_VIDEO_FILL)
+            }
+
+            publisherViewContainer.addView(publisher?.view)
+
+            (publisher?.view as? GLSurfaceView)?.let {
+                it.setZOrderOnTop(true)
+            }
+
+            session.publish(publisher)
+        }
+
+        override fun onDisconnected(session: Session) {
+            Log.d(TAG, "onDisconnected: disconnected from session " + session.sessionId)
+            this@MainActivity.session = null
+        }
+
+        override fun onError(session: Session, opentokError: OpentokError) {
+            finishWithMessage("Session error: " + opentokError.message)
+        }
+
+        override fun onStreamReceived(session: Session, stream: Stream) {
+            Log.d(TAG, "onStreamReceived: New stream " + stream.streamId + " in session " + session.sessionId)
+
+            if (subscriber != null) {
+                return
+            }
+            subscribeToStream(stream)
+        }
+
+        override fun onStreamDropped(session: Session, stream: Stream) {
+            Log.d(TAG, "onStreamDropped: Stream " + stream.streamId + " dropped from session " + session.sessionId)
+
+            if (subscriber == null) {
+                return
+            }
+            if (subscriber?.stream == stream) {
+                subscriberViewContainer.removeView(subscriber?.view)
+                subscriber = null
+            }
+        }
+    }
+    private val videoListener: VideoListener = object : VideoListener {
+        override fun onVideoDataReceived(subscriberKit: SubscriberKit) {
+            subscriber?.setStyle(BaseVideoRenderer.STYLE_VIDEO_SCALE, BaseVideoRenderer.STYLE_VIDEO_FILL)
+            subscriberViewContainer.addView(subscriber?.view)
+        }
+
+        override fun onVideoDisabled(subscriberKit: SubscriberKit, s: String) {}
+        override fun onVideoEnabled(subscriberKit: SubscriberKit, s: String) {}
+        override fun onVideoDisableWarning(subscriberKit: SubscriberKit) {}
+        override fun onVideoDisableWarningLifted(subscriberKit: SubscriberKit) {}
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        if (!isValid) {
+
```

Comment:
remove uneccsary empty space

Reply:
LGTM - this is default AS formatting - which space are you referring to?

Reply:
agreed